_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+Niestety przy rozwijaniu aplikacji takie podejście staje się bardzo nieefektywne, zarówno pod względem organizacji, jak i, z czasem, również wydajności. Zdecydowanie lepiej jest podzielić store na tak zwane "slice'y", czyli wycinki danych, na przykład dzieląc store według funkcjonalności aplikacji. Zobaczmy takie podejście na przykładzie:
+
+```typescript
+export interface State {
+    activitiesState: ActivityItemsState;
+}
+
+export interface ActivityItemsState {
+    activities: ActivityItemModel[];
+}
+
+export interface ActivityItemModel {
+    name: string;
+    participants: number;
+}
+
+```
+
+Przejdźmy po powyższych interfejsach po kolei. Pierwszy z nich mówi, że spodziewamy się mieć jeden wycinek danych (slice) typu `ActivityItemsState`, w którym będą przechowywane wszystkie informacje dotyczące aktywności dodanych do aplikacji. Następnie definiujemy, co tak naprawdę znajduje się w tym wycinku danych - jest to tablica obiektów typu `ActivityItemModel`, czyli informacji o różnych aktywnościach. Ostatni interfejs to już wyłącznie definicja modelu biznesowego - w naszej aplikacji będziemy mieli do czynienia z nazwą czynności oraz możliwą liczbą jej uczestników.
+
+Taki podział store zapewnia bardzo łatwą jego rozszerzalność, gdyż w przypadku dodania nowej funkcjonalności wystarczy dopisać nowy wycinek danych do store. W bardzo dużych aplikacjach, gdzie obiekt stanu aplikacji jest już dość potężny, to podejście ma jeszcze jedną zaletę - pozwala wykorzystać _lazy loading_, to znaczy wczytywać tylko konkretne wycinki store, potrzebne do aktualnie przetwarzanej części aplikacji. 
+
+### Akcje
+Potrzebujemy teraz możliwości wywoływania akcji, żeby powodować zmiany w store. Aplikacja będzie przewidywała dwa zdarzenia: wystosowanie żądania do API po pomysł na jakieś zajęcie oraz dodanie aktywności do listy. Wygląda to następująco:
+
+```typescript
+export const addActivityType = 'Add activity';
+export const getActivityType = 'Get activity';
+
+export const addActivity = createAction(addActivityType, props<ActivityItemModel>());
+export const getActivity = createAction(getActivityType);
+```
+
+Każda akcja musi mieć zdefiniowany swój typ. NgRx używa do tego tzw. _literal type_, czyli po prostu używa typu string do identyfikacji akcji (u nas odpowiednio `Add activity` oraz `Get activity`. W nazywaniu akcji mamy pełną dowolność. Dodatkowo w przypadku dodawania nowej aktywności do listy musimy ją przekazać w ciele akcji przy pomocy metody `props<>()`. Dzięki wykorzystaniu funkcji `createAction` z biblioteki NgRx tworzenie akcji, jak widać, jest bardzo proste.
+
+### Efekt
```

Comment:
i tutaj dalej dyskusja o nazwach i przepływie.

Proponujesz tak
Wejście na ekran -> GetActivities -> ActivityEffect -> Kazali pobrać -> Zlecenie http ...

A ja myślę o
Zdarzenie nawigacji -> ActicityEffect -> Oho! Tu jet potrzebne -> Zlecenie http ...

Jak wtedy wygląda ta elastyczność i skalowlność?
W moim dopisujesz komponent, a potrzeba zasilenia store jest w kodzie logiki. W Twoim trzeba modyfikować komponenty prezentacji i dopisywać w nich zlecenie pobierania.

Teraz wyobraź sobie, że "pobieramy gdy starsze niż X ale gdy coś tam" i zastanówmy się jak to się skaluje? :)